sherlock-webhook-proxy/pkg/handler.go

```diff
@@ -0,0 +1,248 @@
+	"github.com/go-playground/webhooks/v6/github"
+	"hash"
+	"io"
+	"log"
+	"net/http"
+	"net/url"
+	"os"
+	"strings"
+	"time"
+)
+
+const (
+	// sherlockUrlEnvVar should be set like "https://sherlock.dsp-devops.broadinstitute.org" or "http://localhost:8080"
+	sherlockUrlEnvVar = "SHERLOCK_URL"
+	// iapTokenOverrideEnvVar overrides IAP-token generating behavior, to allow local dev
+	iapTokenOverrideEnvVar = "IAP_TOKEN"
+	// iapAudienceEnvVar should be set to the Client ID of the IAP OAuth credentials
+	iapAudienceEnvVar = "IAP_AUDIENCE"
+	// githubWebhookSecretEnvVar should be set to the secret set in the GitHub Webhook config
+	githubWebhookSecretEnvVar = "GITHUB_WEBHOOK_SECRET"
+	// idTokenUrl is the URL to use to get an IAP ID token. Will be used with ?audience=${iapAudienceEnvVar}
+	idTokenUrl = "http://metadata.google.internal/computeMetadata/v1/instance/service-accounts/default/identity"
+)
+
+var (
+	_mac       hash.Hash
+	_hook      *github.Webhook
+	_transport *httptransport.Runtime
+)
+
+// init does first-time setup
+func init() {
+	sherlockUrl, present := os.LookupEnv(sherlockUrlEnvVar)
+	if !present {
+		log.Fatalf("os.LookupEnv(%s): present=false\n", sherlockUrlEnvVar)
+	} else if sherlockUrl == "" {
+		log.Fatalf("os.LookupEnv(%s): sherlockUrl=''\n", sherlockUrlEnvVar)
+	}
+
+	parsedSherlockUrl, err := url.Parse(sherlockUrl)
+	if err != nil {
+		log.Fatalf("url.Parse(%s): %v\n", sherlockUrl, err)
+	}
+	hostname := parsedSherlockUrl.Hostname()
+	if parsedSherlockUrl.Port() != "" {
+		hostname += ":"
+		hostname += parsedSherlockUrl.Port()
+	}
+	_transport = httptransport.New(hostname, "", []string{parsedSherlockUrl.Scheme})
+
+	if _, present := os.LookupEnv(iapTokenOverrideEnvVar); !present {
+		iapAudience, present := os.LookupEnv(iapAudienceEnvVar)
+		if !present {
+			log.Fatalf("os.LookupEnv(%s): present=false\n", iapAudienceEnvVar)
+		} else if iapAudience == "" {
+			log.Fatalf("os.LookupEnv(%s): sherlockUrl=''\n", iapAudienceEnvVar)
+		}
+	}
+
+	secret, present := os.LookupEnv(githubWebhookSecretEnvVar)
+	if !present {
+		log.Fatalf("os.LookupEnv(%s): present=false\n", githubWebhookSecretEnvVar)
+	} else if secret == "" {
+		log.Fatalf("os.LookupEnv(%s): secret=''\n", githubWebhookSecretEnvVar)
+	}
+	_mac = hmac.New(sha256.New, []byte(secret))
+
+	hook, err := github.New(github.Options.Secret(secret))
+	if err != nil {
+		log.Fatalf("github.New: %v\n", err)
+	}
+	_hook = hook
+}
+
+// HandleWebhook is what actually does the handling, running once per request
+func HandleWebhook(w http.ResponseWriter, r *http.Request) {
+	switch r.RequestURI {
+	case "/webhook":
+		// Require that the signature header is present
+		signature := strings.TrimPrefix(r.Header.Get("X-Hub-Signature-256"), "sha256=")
+		if signature == "" {
+			w.WriteHeader(http.StatusUnauthorized)
+			log.Printf("request was missing X-Hub-Signature-256 header\n")
+			return
+		}
+
+		// As r.Body is read, additionally synchronously write it into _mac
+		_mac.Reset()
```

Comment:
does cloud functions guarantee that only one request is handled at a time? otherwise keeping this in a package-level variable doesn't seem concurrency safe 🤔

Reply:
It does! With cloud functions gen2 it would be possible for us to override that, to make it handle requests concurrently, but by default they're one at a time. It does keep instances alive, which is why I'm using `init` and package-level stuff to make that run faster after cold-start